File: scala/scalafind/src/main/scala/scalafind/Finder.scala
package scalafind

import scalafind.FileType.FileType
import scalafind.FileUtil.getExtension
import scalafind.FindError.{INVALID_RANGE_MINDEPTH_MAXDEPTH, INVALID_RANGE_MINLASTMOD_MAXLASTMOD,
  INVALID_RANGE_MINSIZE_MAXSIZE, STARTPATH_DOES_NOT_MATCH, STARTPATH_NOT_DEFINED,
  STARTPATH_NOT_FOUND, STARTPATH_NOT_READABLE}

import java.io.IOException
import java.nio.file.attribute.{BasicFileAttributes, FileTime}
import java.nio.file.{Files, Path}
import java.time.{Instant, LocalDateTime, ZoneOffset}
import scala.collection.mutable
import scala.util.matching.Regex

class Finder (settings: FindSettings) {
  import Finder.*

  private def validateSettings(): Unit = {
    if (settings.paths.isEmpty) {
      throw new FindException(STARTPATH_NOT_DEFINED.toString)
    }
    settings.paths.foreach(path =>
      val p =
        if (Files.exists(path)) {
          path
        } else {
          val expandedPath = FileUtil.expandPath(path)
          if (Files.exists(expandedPath)) {
            expandedPath
          } else {
            throw new FindException(STARTPATH_NOT_FOUND.toString)
          }
        }
      if (!Files.isReadable(p)) {
        throw new FindException(STARTPATH_NOT_READABLE.toString)
      }
      if (Files.isSymbolicLink(p)) {
        if (!settings.followSymlinks) {
          throw new FindException(STARTPATH_DOES_NOT_MATCH.toString)
        }
      } else if (Files.isDirectory(p)) {
        if (!isTraversableDirPath(p)) {
          throw new FindException(STARTPATH_DOES_NOT_MATCH.toString)
        }
      } else if (Files.isRegularFile(p)) {
        if (filterToFileResult(p).isEmpty) {
          throw new FindException(STARTPATH_DOES_NOT_MATCH.toString)
        }
      } else {
        // TODO: start path is unknown/invalid type
        throw new FindException(STARTPATH_DOES_NOT_MATCH.toString)
      }
    )
    if (settings.maxDepth > -1 && settings.minDepth > settings.maxDepth) {
      throw new FindException(INVALID_RANGE_MINDEPTH_MAXDEPTH.toString)
    }
    if (compareOptionLocalDateTimes(settings.maxLastMod, settings.minLastMod) < 0) {
      throw new FindException(INVALID_RANGE_MINLASTMOD_MAXLASTMOD.toString)
    }
    if (settings.maxSize > 0 && settings.minSize > settings.maxSize) {
      throw new FindException(INVALID_RANGE_MINSIZE_MAXSIZE.toString)
    }
  }
  validateSettings()

  private def isMatchingPathBySymlink(path: Path): Boolean = {
    settings.followSymlinks || !Files.isSymbolicLink(path)
  }

  private def isMatchingDirPathByHidden(dirPath: Path): Boolean = {
    settings.includeHidden || !FileUtil.isHiddenPath(dirPath)
  }

  private def isMatchingDirPathByInPatterns(dirPath: Path): Boolean = {
    emptyOrAnyMatchesAnyPattern(FileUtil.splitPath(dirPath), settings.inDirPatterns)
  }

  private def isMatchingDirPathByOutPatterns(dirPath: Path): Boolean = {
    emptyOrNotMatchesAnyPattern(dirPath.toString, settings.outDirPatterns)
  }

  def isTraversableDirPath(dirPath: Path): Boolean = {
    isMatchingPathBySymlink(dirPath)
      && isMatchingDirPathByHidden(dirPath)
      && isMatchingDirPathByOutPatterns(dirPath)
  }

  def isMatchingDirPath(dirPath: Path): Boolean = {
    isMatchingPathBySymlink(dirPath)
      && isMatchingDirPathByHidden(dirPath)
      && isMatchingDirPathByInPatterns(dirPath)
      && isMatchingDirPathByOutPatterns(dirPath)
  }

  def isNullOrMatchingDirPath(dirPath: Path): Boolean = {
    dirPath== null || dirPath.toString.isEmpty || isMatchingDirPath(dirPath)
  }

  private def isMatchingFileNameByHidden(fileName: String): Boolean = {
    settings.includeHidden || !FileUtil.isHiddenName(fileName)
  }

  def isMatchingArchiveExtension(ext: String): Boolean = {
    emptyOrMatchesAnyString(ext, settings.inArchiveExtensions)
      && emptyOrNotMatchesAnyString(ext, settings.outArchiveExtensions)
  }

  def isMatchingArchiveExtensionForFilePath(filePath: Path): Boolean = {
    if (settings.inArchiveExtensions.nonEmpty || settings.outArchiveExtensions.nonEmpty) {
      val ext = getExtension(filePath.getFileName.toString)
      isMatchingArchiveExtension(ext)
    } else {
      true
    }
  }

  def isMatchingArchiveFileName(fileName: String): Boolean = {
    emptyOrMatchesAnyPattern(fileName, settings.inArchiveFilePatterns)
      && emptyOrNotMatchesAnyPattern(fileName, settings.outArchiveFilePatterns)
  }

  def isMatchingArchiveFileNameForFilePath(filePath: Path): Boolean = {
    if (settings.inArchiveFilePatterns.nonEmpty || settings.outArchiveFilePatterns.nonEmpty) {
      isMatchingArchiveFileName(filePath.getFileName.toString)
    } else {
      true
    }
  }

  def isMatchingArchiveFilePath(filePath: Path): Boolean = {
    isMatchingArchiveExtensionForFilePath(filePath)
      && isMatchingArchiveFileNameForFilePath(filePath)
  }

  def isMatchingArchiveFileResult(fr: FileResult): Boolean = {
    isMatchingArchiveFilePath(fr.path)
  }

  private def isMatchingExtension(ext: String): Boolean = {
    emptyOrMatchesAnyString(ext, settings.inExtensions)
      && emptyOrNotMatchesAnyString(ext, settings.outExtensions)
  }

  private def isMatchingExtensionForFilePath(filePath: Path): Boolean = {
    if (settings.inExtensions.nonEmpty || settings.outExtensions.nonEmpty) {
      isMatchingExtension(getExtension(filePath.getFileName.toString))
    } else {
      true
    }
  }

  private def isMatchingFileName(fileName: String): Boolean = {
    emptyOrMatchesAnyPattern(fileName, settings.inFilePatterns)
      && emptyOrNotMatchesAnyPattern(fileName, settings.outFilePatterns)
  }

  private def isMatchingFileNameForFilePath(filePath: Path): Boolean = {
    if (settings.inFilePatterns.nonEmpty || settings.outFilePatterns.nonEmpty) {
      isMatchingFileName(filePath.getFileName.toString)
    } else {
      true
    }
  }

  private def isMatchingFilePath(filePath: Path): Boolean = {
    // We assume that isNullOrMatchingDirPath(filePath.getParent()) has already been called
    isMatchingExtensionForFilePath(filePath)
      && isMatchingFileNameForFilePath(filePath)
  }

  private def isMatchingFileType(fileType: FileType): Boolean = {
    emptyOrMatchesAnyFileType(fileType, settings.inFileTypes)
      && emptyOrNotMatchesAnyFileType(fileType, settings.outFileTypes)
  }

  private def isMatchingSize(fileSize: Long): Boolean = {
    (settings.maxSize <= 0 || fileSize <= settings.maxSize)
      && (settings.minSize <= 0 || fileSize >= settings.minSize)
  }

  private def isMatchingLastMod(lastMod: Option[Instant]): Boolean = {
    // TODO: lastMod is not needed if maxLastMod and minLastMod are not set, so can avoid getting last mod time for files if not needed
    (settings.maxLastMod, settings.minLastMod) match {
      case (Some(max), Some(min)) =>
        lastMod.nonEmpty &&
        lastMod.get.compareTo(max.toInstant(ZoneOffset.UTC)) <= 0 &&
          lastMod.get.compareTo(min.toInstant(ZoneOffset.UTC)) >= 0
      case (Some(max), None) =>
        lastMod.nonEmpty &&
        lastMod.get.compareTo(max.toInstant(ZoneOffset.UTC)) <= 0
      case (None, Some(min)) =>
        lastMod.nonEmpty &&
        lastMod.get.compareTo(min.toInstant(ZoneOffset.UTC)) >= 0
      case (None, None) =>
        true
    }
  }

  def isMatchingFileResult(fr: FileResult): Boolean = {
    isMatchingFilePath(fr.path)
      && isMatchingFileType(fr.fileType)
      && isMatchingSize(fr.fileSize)
      && isMatchingLastMod(fr.lastMod.map(_.toInstant))
  }

  private def getFileSizeAndLastMod(filePath: Path): (Long, Option[FileTime]) = {
    if (settings.needLastMod || settings.needSize) {
      try {
        val stat = Files.readAttributes(filePath, classOf[BasicFileAttributes])
        val size = if (settings.needSize) stat.size() else 0L
        val lastMod = if (settings.needLastMod) Some(stat.lastModifiedTime()) else None
        (size, lastMod)
      } catch {
        case _: IOException => (0L, None)
      }
    } else {
      (0L, None)
    }
  }

  private def filterArchiveFilePathToFileResult(filePath: Path, fileType: FileType): Option[FileResult] = {
    if (!settings.includeArchives && !settings.archivesOnly) {
      None
    } else if (!isMatchingArchiveFilePath(filePath)) {
      None
    } else {
      Some(new FileResult(filePath, fileType, 0L, None))
    }
  }

  private def filterRegularFilePathToFileResult(filePath: Path, fileType: FileType): Option[FileResult] = {
    if (settings.archivesOnly) {
      None
    } else if (!isMatchingFilePath(filePath) || !isMatchingFileType(fileType)) {
      None
    } else {
      val (fileSize, lastMod) = getFileSizeAndLastMod(filePath)
      if (!isMatchingSize(fileSize) || !isMatchingLastMod(lastMod.map(_.toInstant))) {
        None
      } else {
        Some(new FileResult(filePath, fileType, fileSize, lastMod))
      }
    }
  }

  def filterToFileResult(filePath: Path): Option[FileResult] = {
    if (!isNullOrMatchingDirPath(filePath.getParent)
      || !isMatchingFileNameByHidden(filePath.getFileName.toString)) {
      None
    } else {
      val fileType = FileTypes.getFileType(filePath)
      if (fileType == FileType.Archive) {
        filterArchiveFilePathToFileResult(filePath, fileType)
      } else {
        filterRegularFilePathToFileResult(filePath, fileType)
      }
    }
  }

  private final def recFindPath(path: Path, minDepth: Int, maxDepth: Int, currentDepth: Int): Seq[FileResult] = {
    val recurse =
      if (currentDepth == maxDepth) {
        RecursionType.NoRecurse
      } else if (maxDepth > -1 && currentDepth > maxDepth) {
        RecursionType.Skip
      } else {
        RecursionType.Recurse
      }
    if (recurse == RecursionType.Skip) {
      Seq.empty[FileResult]
    } else {
      val pathDirs = mutable.ArrayBuffer.empty[Path]
      val pathResults = mutable.ArrayBuffer.empty[FileResult]
      try {
        val subPathStream: java.nio.file.DirectoryStream[Path] = Files.newDirectoryStream(path)
        val iterator = subPathStream.iterator()
        while (iterator.hasNext) {
          val subPath = iterator.next
          if (!Files.isSymbolicLink(subPath) || settings.followSymlinks) {
            if (Files.isDirectory(subPath) && recurse == RecursionType.Recurse && isTraversableDirPath(subPath)) {
              pathDirs += subPath
            } else {
              if (Files.isRegularFile(subPath) && (minDepth < 0 || currentDepth >= minDepth)) {
                val optFileResult: Option[FileResult] = filterToFileResult(subPath)
                optFileResult.foreach(pathResults += _)
              }
            }
          }
        }
      } catch {
        case e: IOException =>
          e.printStackTrace()
      }
      Seq.empty ++ pathDirs.flatMap(recFindPath(_, minDepth, maxDepth, currentDepth + 1))
    }
  }

  private final def findPath(path: Path): Seq[FileResult] = {
    val p =
      if (Files.exists(path)) {
        path
      } else {
        val expandedPath = FileUtil.expandPath(path)
        if (Files.exists(expandedPath)) {
          expandedPath
        } else {
          throw new FindException(STARTPATH_NOT_FOUND.toString)
        }
      }
    if (Files.isSymbolicLink(p) && !settings.followSymlinks) {
      throw new FindException(STARTPATH_DOES_NOT_MATCH.toString)
    }
    if (Files.isDirectory(p)) {
      if (settings.maxDepth == 0) {
        Seq.empty[FileResult]
      } else {
        if (isTraversableDirPath(p)) {
          val maxDepth = if (settings.recursive) settings.maxDepth else 1
          recFindPath(p, settings.minDepth, maxDepth, 1)
        } else {
          throw new FindException(STARTPATH_DOES_NOT_MATCH.toString)
        }
      }
    } else if (Files.isRegularFile(p)) {
      if (settings.minDepth > 0) {
        Seq.empty[FileResult]
      } else {
        filterToFileResult(p) match {
          case Some(fileResult) =>
            Seq(fileResult)
          case None =>
            throw new FindException(STARTPATH_DOES_NOT_MATCH.toString)
        }
      }
    } else {
      throw new FindException(STARTPATH_DOES_NOT_MATCH.toString)
    }
  }

  def find(): Seq[FileResult] = {
    val fileResults = mutable.ArrayBuffer.empty[FileResult]
    settings.paths.foreach { path =>
      fileResults ++= findPath(path)
    }
    if (fileResults.size > 1) {
      val fileResultSorter = new FileResultSorter(settings)
      fileResultSorter.sort(Seq.empty[FileResult] ++ fileResults)
    } else {
      Seq.empty[FileResult] ++ fileResults
    }
  }
}

object Finder {
  private def compareOptionLocalDateTimes(d1: Option[LocalDateTime], d2: Option[LocalDateTime]): Int = {
    if (d1.isEmpty || d2.isEmpty) {
      0
    } else {
      d1.get.compareTo(d2.get)
    }
  }

  private def matchesAnyPattern(s: String, patterns: Set[Regex]): Boolean = {
    patterns exists (_.findFirstMatchIn(s).isDefined)
  }

  private def anyMatchesAnyPattern(strings: Iterable[String], patterns: Set[Regex]): Boolean = {
    strings exists (matchesAnyPattern(_, patterns))
  }

  private def emptyOrMatchesAnyPattern(s: String, patterns: Set[Regex]): Boolean = {
    patterns.isEmpty || patterns.exists (_.findFirstMatchIn(s).isDefined)
  }

  private def emptyOrNotMatchesAnyPattern(s: String, patterns: Set[Regex]): Boolean = {
    patterns.isEmpty || !patterns.exists (_.findFirstMatchIn(s).isDefined)
  }

  private def emptyOrAnyMatchesAnyPattern(strings: Iterable[String], patterns: Set[Regex]): Boolean = {
    patterns.isEmpty || anyMatchesAnyPattern(strings, patterns)
  }

  private def emptyOrNotAnyMatchesAnyPattern(strings: Iterable[String], patterns: Set[Regex]): Boolean = {
    patterns.isEmpty || !anyMatchesAnyPattern(strings, patterns)
  }

  private def emptyOrMatchesAnyString(s: String, stringSet: Set[String]): Boolean = {
    stringSet.isEmpty || stringSet.contains(s)
  }

  private def emptyOrNotMatchesAnyString(s: String, stringSet: Set[String]): Boolean = {
    stringSet.isEmpty || !stringSet.contains(s)
  }

  private def emptyOrMatchesAnyFileType(fileType: FileType, fileTypeSet: Set[FileType]): Boolean = {
    fileTypeSet.isEmpty || fileTypeSet.contains(fileType)
  }

  private def emptyOrNotMatchesAnyFileType(fileType: FileType, fileTypeSet: Set[FileType]): Boolean = {
    fileTypeSet.isEmpty || !fileTypeSet.contains(fileType)
  }

  private def getMatchingDirs(fileResults: Seq[FileResult]): Seq[Path] = {
    fileResults
      .map(fr => FileUtil.pathOrCurrent(fr.path))
      .distinct
      .sortWith(_.toString < _.toString)
  }

  def printMatchingDirs(fileResults: Seq[FileResult], formatter: FileResultFormatter): Unit = {
    val dirs = getMatchingDirs(fileResults)
    if (dirs.nonEmpty) {
      Common.log("\nMatching directories (%d):".format(dirs.length))
      dirs.foreach(d => Common.log(formatter.formatDirPath(d)))
    } else {
      Common.log("\nMatching directories: 0")
    }
  }

  def printMatchingFiles(fileResults: Seq[FileResult], formatter: FileResultFormatter): Unit = {
    if (fileResults.nonEmpty) {
      Common.log("\nMatching files (%d):".format(fileResults.length))
      fileResults.foreach(fr => Common.log(formatter.formatFileResult(fr)))
    } else {
      Common.log("\nMatching files: 0")
    }
  }
}

object RecursionType extends Enumeration {
  type RecursionType = Value
  val Skip, NoRecurse, Recurse = Value
}

File: kotlin/ktfind/lib/src/main/kotlin/ktfind/Finder.kt
package ktfind

import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.attribute.BasicFileAttributes
import java.nio.file.attribute.FileTime
import java.time.Instant
import java.time.ZoneOffset
import kotlin.io.path.extension
import kotlin.io.path.isSymbolicLink
import kotlin.io.path.name

/**
 * @author cary on 7/23/16.
 */
class Finder(val settings: FindSettings) {
    private val fileTypes: FileTypes = FileTypes()

    init {
        validateSettings(settings)
    }

    private fun validateSettings(settings: FindSettings) {
        if (settings.paths.isEmpty()) {
            throw FindException(FindError.STARTPATH_NOT_DEFINED.message)
        }
        for (path in settings.paths) {
            var p = path
            if (!Files.exists(p)) {
                p = FileUtil.expandPath(p)
                if (!Files.exists(p)) {
                    throw FindException(FindError.STARTPATH_NOT_FOUND.message)
                }
            }
            if (!Files.isReadable(p)) {
                throw FindException(FindError.STARTPATH_NOT_READABLE.message)
            }
            if (Files.isSymbolicLink(p)) {
                if (!settings.followSymlinks) {
                    throw FindException(FindError.STARTPATH_DOES_NOT_MATCH_FIND_SETTINGS.message)
                }
            } else if (Files.isDirectory(p)) {
                if (!isTraversableDirPath(p)) {
                    throw FindException(FindError.STARTPATH_DOES_NOT_MATCH_FIND_SETTINGS.message)
                }
            } else if (Files.isRegularFile(p)) {
                if (filterToFileResult(p) == null) {
                    throw FindException(FindError.STARTPATH_DOES_NOT_MATCH_FIND_SETTINGS.message)
                }
            } else {
                // TODO: start path is unknown/invalid type
                throw FindException(FindError.STARTPATH_DOES_NOT_MATCH_FIND_SETTINGS.message)
            }
        }
        if (settings.maxDepth > -1 && settings.maxDepth < settings.minDepth) {
            throw FindException(FindError.INVALID_RANGE_FOR_MINDEPTH_AND_MAXDEPTH.message)
        }
        if (settings.maxLastMod != null && settings.minLastMod != null
            && settings.maxLastMod < settings.minLastMod) {
            throw FindException(FindError.INVALID_RANGE_FOR_MINLASTMOD_AND_MAXLASTMOD.message)
        }
        if (settings.maxSize > 0 && settings.maxSize < settings.minSize) {
            throw FindException(FindError.INVALID_RANGE_FOR_MINSIZE_AND_MAXSIZE.message)
        }
    }

    private fun matchesAnyPattern(s: String, patternSet: Set<Regex>): Boolean {
        return patternSet.any { p -> p.containsMatchIn(s) }
    }

    private fun anyMatchesAnyPattern(
        sList: List<String>,
        patternSet: Set<Regex>
    ): Boolean {
        return sList.any { s -> matchesAnyPattern(s, patternSet) }
    }

    private fun emptyOrMatchesAnyPattern(s: String, patternSet: Set<Regex>): Boolean {
        return patternSet.isEmpty() || matchesAnyPattern(s, patternSet)
    }

    private fun emptyOrNotMatchesAnyPattern(s: String, patternSet: Set<Regex>): Boolean {
        return patternSet.isEmpty() || !matchesAnyPattern(s, patternSet)
    }

    private fun emptyOrAnyMatchesAnyPattern(sList: List<String>, patternSet: Set<Regex>): Boolean {
        return patternSet.isEmpty() || anyMatchesAnyPattern(sList, patternSet)
    }

    private fun emptyOrNotAnyMatchesAnyPattern(sList: List<String>, patternSet: Set<Regex>): Boolean {
        return patternSet.isEmpty() || !anyMatchesAnyPattern(sList, patternSet)
    }

    private fun emptyOrMatchesAnyString(s: String, stringSet: Set<String>): Boolean {
        return stringSet.isEmpty() || stringSet.contains(s)
    }

    private fun emptyOrNotMatchesAnyString(s: String, stringSet: Set<String>): Boolean {
        return stringSet.isEmpty() || !stringSet.contains(s)
    }

    private fun emptyOrMatchesAnyFileType(fileType: FileType, fileTypeSet: Set<FileType>): Boolean {
        return fileTypeSet.isEmpty() || fileTypeSet.contains(fileType)
    }

    private fun emptyOrNotMatchesAnyFileType(fileType: FileType, fileTypeSet: Set<FileType>): Boolean {
        return fileTypeSet.isEmpty() || !fileTypeSet.contains(fileType)
    }

    fun isMatchingPathBySymlink(path: Path): Boolean {
        return settings.followSymlinks || !path.isSymbolicLink()
    }

    fun isMatchingDirPathByHidden(dirPath: Path): Boolean {
        return settings.includeHidden || !FileUtil.isHiddenPath(dirPath)
    }

    fun isMatchingDirPathByInPatterns(dirPath: Path): Boolean {
        val pathElems = FileUtil.splitPath(dirPath)
        return emptyOrAnyMatchesAnyPattern(pathElems, settings.inDirPatterns)
    }

    fun isMatchingDirPathByOutPatterns(dirPath: Path): Boolean {
        val pathElems = FileUtil.splitPath(dirPath)
        return emptyOrNotAnyMatchesAnyPattern(pathElems, settings.outDirPatterns)
    }

    fun isTraversableDirPath(dirPath: Path): Boolean {
        return isMatchingDirPathByHidden(dirPath)
                && isMatchingDirPathByOutPatterns(dirPath)
    }

    fun isMatchingDirPath(dirPath: Path): Boolean {
        return isMatchingDirPathByHidden(dirPath)
                && isMatchingDirPathByInPatterns(dirPath)
                && isMatchingDirPathByOutPatterns(dirPath)
    }

    fun isNullOrMatchingDirPath(dirPath: Path?): Boolean {
        // null or empty path is a match
        if (dirPath == null || dirPath.toString().isEmpty()) {
            return true
        }
        return isMatchingDirPath(dirPath)
    }

    fun isMatchingFileNameByHidden(fileName: String): Boolean {
        return settings.includeHidden || !FileUtil.isHiddenName(fileName)
    }

    fun isMatchingArchiveExtension(ext: String): Boolean {
        return emptyOrMatchesAnyString(ext, settings.inArchiveExtensions)
                && emptyOrNotMatchesAnyString(ext, settings.outArchiveExtensions)
    }

    fun isMatchingArchiveExtensionForFilePath(filePath: Path): Boolean {
        if (!settings.inArchiveExtensions.isEmpty() || !settings.outArchiveExtensions.isEmpty()) {
            return isMatchingArchiveExtension(filePath.extension)
        }
        return true
    }

    fun isMatchingArchiveFileName(fileName: String): Boolean {
        return emptyOrMatchesAnyPattern(fileName, settings.inArchiveFilePatterns)
                && emptyOrNotMatchesAnyPattern(fileName, settings.outArchiveFilePatterns)
    }

    fun isMatchingArchiveFileNameForFilePath(filePath: Path): Boolean {
        if (!settings.inArchiveFilePatterns.isEmpty() || !settings.outArchiveFilePatterns.isEmpty()) {
            val fileName = if (filePath.fileName == null) "" else filePath.fileName.toString()
            return isMatchingArchiveFileName(fileName)
        }
        return true
    }

    fun isMatchingArchiveFilePath(filePath: Path): Boolean {
        return isMatchingArchiveExtensionForFilePath(filePath)
                && isMatchingArchiveFileNameForFilePath(filePath)
    }

    fun isMatchingArchiveFileResult(fr: FileResult): Boolean {
        return isNullOrMatchingDirPath(fr.path.parent)
                && isMatchingArchiveFilePath(fr.path)
    }

    fun isMatchingExtension(ext: String): Boolean {
        return emptyOrMatchesAnyString(ext, settings.inExtensions)
                && emptyOrNotMatchesAnyString(ext, settings.outExtensions)
    }

    fun isMatchingExtensionForFilePath(filePath: Path): Boolean {
        if (!settings.inExtensions.isEmpty() || !settings.outExtensions.isEmpty()) {
            return isMatchingExtension(filePath.extension)
        }
        return true
    }

    fun isMatchingFileName(fileName: String): Boolean {
        return emptyOrMatchesAnyPattern(fileName, settings.inFilePatterns)
                && emptyOrNotMatchesAnyPattern(fileName, settings.outFilePatterns)
    }

    fun isMatchingFileNameForFilePath(filePath: Path): Boolean {
        if (!settings.inFilePatterns.isEmpty() || !settings.outFilePatterns.isEmpty()) {
            return isMatchingFileName(filePath.fileName.toString())
        }
        return true
    }

    fun isMatchingFilePath(filePath: Path): Boolean {
        // We assume that isNullOrMatchingDirPath(filePath.getParent()) has already been called
        return isMatchingExtensionForFilePath(filePath)
                && isMatchingFileNameForFilePath(filePath)
    }

    fun isMatchingFileType(fileType: FileType): Boolean {
        return emptyOrMatchesAnyFileType(fileType, settings.inFileTypes)
                && emptyOrNotMatchesAnyFileType(fileType, settings.outFileTypes)
    }

    fun isMatchingFileSize(fileSize: Long): Boolean {
        return ((settings.maxSize <= 0 || fileSize <= settings.maxSize)
                && (settings.minSize <= 0 || fileSize >= settings.minSize))
    }

    fun isMatchingLastMod(lastMod: Instant?): Boolean {
        if (settings.maxLastMod == null && settings.minLastMod == null) {
            return true
        }
        if (lastMod == null) {
            return false
        }
        if (settings.maxLastMod != null && lastMod > settings.maxLastMod.toInstant(ZoneOffset.UTC)) {
            return false
        }
        return ((settings.minLastMod == null)
                || lastMod >= settings.minLastMod.toInstant(ZoneOffset.UTC))
    }

    fun isMatchingLastMod(lastMod: FileTime?): Boolean {
        val lastModInstant = lastMod?.toInstant()
        return isMatchingLastMod(lastModInstant)
    }

    fun isMatchingFileResult(fr: FileResult): Boolean {
        return isNullOrMatchingDirPath(fr.path.parent)
                && isMatchingFilePath(fr.path)
                && isMatchingFileType(fr.fileType)
                && isMatchingFileSize(fr.fileSize)
                && isMatchingLastMod(fr.lastMod?.toInstant())
    }

    fun filterArchiveFilePathToFileResult(filePath: Path, fileType: FileType): FileResult? {
        if (!settings.includeArchives && !settings.archivesOnly) {
            return null
        }
        if (!isMatchingArchiveFilePath(filePath)) {
            return null
        }
        return FileResult(filePath, fileType, 0L, null)
    }

    fun filterRegularFilePathToFileResult(filePath: Path, fileType: FileType): FileResult? {
        if (settings.archivesOnly) {
            return null
        }

        if (!isMatchingFilePath(filePath) || !isMatchingFileType(fileType)) {
            return null
        }

        var fileSize = 0L
        var lastMod: FileTime? = null
        if (needLastMod(settings) || needSize(settings)) {
            try {
                val stat: BasicFileAttributes = Files.readAttributes(filePath, BasicFileAttributes::class.java)
                if (needSize(settings)) fileSize = stat.size()
                if (needLastMod(settings)) lastMod = stat.lastModifiedTime()
            } catch (e: Exception) {
                logError(e.message!!)
                return null
            }
        }

        if (!isMatchingFileSize(fileSize) || !isMatchingLastMod(lastMod)) {
            return null
        }

        return FileResult(filePath, fileType, fileSize, lastMod)
    }

    fun filterToFileResult(filePath: Path): FileResult? {
        if (!isNullOrMatchingDirPath(filePath.parent)
            || !isMatchingFileNameByHidden(filePath.fileName.toString())) {
            return null
        }

        val fileType = fileTypes.getFileType(filePath)
        if (fileType == FileType.ARCHIVE) {
            return filterArchiveFilePathToFileResult(filePath, fileType)
        }
        return filterRegularFilePathToFileResult(filePath, fileType)
    }

    private fun recFindPath(path: Path, minDepth: Int, maxDepth: Int, currentDepth: Int): List<FileResult> {
        var recurse = true
        if (currentDepth == maxDepth) {
            recurse = false
        } else if (maxDepth > -1 && currentDepth > maxDepth) {
            return emptyList()
        }
        val pathDirs: MutableList<Path> = mutableListOf()
        val pathResults: MutableList<FileResult> = mutableListOf()
        try {
            Files.newDirectoryStream(path).use { stream ->
                stream.forEach {
                    if (!Files.isSymbolicLink(it) || settings.followSymlinks) {
                        if (Files.isDirectory(it) && recurse && isTraversableDirPath(it)) {
                            pathDirs.add(it)
                        } else if (Files.isRegularFile(it) && (minDepth < 0 || currentDepth >= minDepth)) {
                            val fr = filterToFileResult(it)
                            if (fr != null) {
                                pathResults.add(fr)
                            }
                        }
                    }
                }
            }
        } catch (_: Exception) {
            return emptyList()
        }

        pathDirs.forEach {
            pathResults.addAll(recFindPath(it, minDepth, maxDepth, currentDepth + 1))
        }

        return pathResults
    }

    private fun findPath(path: Path): List<FileResult> {
        var p = path
        if (!Files.exists(path)) {
            p = FileUtil.expandPath(path)
            if (!Files.exists(p)) {
                throw FindException(FindError.STARTPATH_NOT_FOUND.message)
            }
        }
        if (Files.isSymbolicLink(p) && !settings.followSymlinks) {
            throw FindException(FindError.STARTPATH_DOES_NOT_MATCH_FIND_SETTINGS.message)
        }
        if (Files.isDirectory(p)) {
            // if max_depth is zero, we can skip since a directory cannot be a result
            if (settings.maxDepth == 0) {
                return emptyList()
            }
            if (isTraversableDirPath(p)) {
                val maxDepth = if (settings.recursive) settings.maxDepth else 1
                return recFindPath(p, settings.minDepth, maxDepth, 1)
            } else {
                throw FindException(FindError.STARTPATH_DOES_NOT_MATCH_FIND_SETTINGS.message)
            }
        } else if (Files.isRegularFile(p)) {
            // if min_depth > zero, we can skip since the file is at depth zero
            if (settings.minDepth > 0) {
                return emptyList()
            }
            val fr = filterToFileResult(p)
            if (fr != null) {
                return listOf(fr)
            } else {
                throw FindException(FindError.STARTPATH_DOES_NOT_MATCH_FIND_SETTINGS.message)
            }
        } else {
            throw FindException(FindError.STARTPATH_DOES_NOT_MATCH_FIND_SETTINGS.message)
        }
    }

    private fun findAsync(fileResults: MutableList<FileResult>): Unit = runBlocking {
        for (path in settings.paths) {
            launch {
                fileResults.addAll(findPath(path))
            }
        }
    }

    fun find(): List<FileResult> {
        val fileResults: MutableList<FileResult> = mutableListOf()
        if (settings.paths.size > 1) {
            findAsync(fileResults)
        } else {
            fileResults.addAll(findPath(settings.paths.first()))
        }
        if (fileResults.size > 1) {
            val fileResultSorter = FileResultSorter(settings)
            return fileResultSorter.sort(fileResults.toList())
        }
        return fileResults.toList()
    }

    fun printMatchingDirs(fileResults: List<FileResult>, formatter: FileResultFormatter) {
        val dirs = fileResults.mapNotNull { f -> f.path.parent }.distinct().sorted()
        if (dirs.isEmpty()) {
            log("\nMatching directories: 0")
        } else {
            log("\nMatching directories (${dirs.size}):")
            for (d in dirs) {
                log(formatter.formatDirPath(d))
            }
        }
    }

    fun printMatchingFiles(fileResults: List<FileResult>, formatter: FileResultFormatter) {
        if (fileResults.isEmpty()) {
            log("\nMatching files: 0")
        } else {
            log("\nMatching files (${fileResults.size}):")
            for (fr in fileResults) {
                log(formatter.formatFileResult(fr))
            }
        }
    }
}
